### PR TITLE
drivers: i2c: added support for Andes atciic100 driver

### PR DESCRIPTION
**Pull request by jimmyzhe**
Added support for Andes atciic100 I2C driver, tested on adp_xc7k_ae350 board.

### CODE REVIEW

File: drivers/i2c/i2c_andes_atciic100.c [hunk 1]
```diff
@@ -0,0 +1,983 @@
...
+	case 0x3:
+		i2c_info->fifo_depth = 16;
+		break;
+	}
+
+	/*
+	 * I2C setting: slave mode(default), standard speed
+	 * 7-bit, CPU mode
+	 */
+	OUTWORD(I2C_SET(dev), 0x0);
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.c:79]: you can likely use the sys_write32/sys_read32 in place of outword/inword macros here

File: drivers/i2c/i2c_andes_atciic100.c [hunk 2]
```diff
@@ -0,0 +1,983 @@
...
+	default:
+		return -EIO;
+	}
+
+	*dev_config = config;
+
+	return 0;
+}
+
+static int i2c_atciic100_transfer(const struct device *dev,
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.c:198]: I would prefer to see a single exit from this function with k_sem_give there avoiding the potential for missing a k_sem_give as the code evolves.

File: drivers/i2c/i2c_andes_atciic100.c [hunk 3]
```diff
@@ -0,0 +1,983 @@
...
+	struct i2c_msg *msgs, uint8_t num_msgs, uint16_t addr)
+{
+	const struct i2c_atciic100_device_config *dev_cfg = DEV_I2C_CFG(dev);
+	struct _i2c_info *i2c_info = dev_cfg->i2c_info;
+	uint8_t burst_write_buf[512] = {0};
+	uint8_t burst_write_len = 0;
+	uint8_t count = 0;
+	uint8_t status = 0;
+
+	if (k_sem_take(&i2c_info->i2c_transfer_sem, K_MSEC(1000))) {
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.c:208]: The timeout here isn't useful in any other place, the API doesn't support timeouts at the moment and is a general issue. Instead use K_FOREVER.

File: drivers/i2c/i2c_andes_atciic100.c [hunk 4]
```diff
@@ -0,0 +1,983 @@
...
+					return status;
+				}
+				status = i2c_atciic100_master_receive(dev,
+					addr, msgs[1].buf, msgs[1].len, true);
+			} else {
+				burst_write_len = msgs[0].len + msgs[1].len;
+
+				if (burst_write_len > MAX_XFER_SZ) {
+					k_sem_give(&i2c_info->i2c_transfer_sem);
+					return DRIVER_ERROR_UNSUPPORTED;
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.c:237]: -EINVAL is fine

File: drivers/i2c/i2c_andes_atciic100.c [hunk 5]
```diff
@@ -0,0 +1,983 @@
...
+	} else {
+		if (num_msgs == 1) {
+			if ((msgs->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+				status = i2c_atciic100_slave_send(dev, msgs->buf, msgs->len);
+			} else {
+				status = i2c_atciic100_slave_receive(dev, msgs->buf, msgs->len);
+			}
+		} else {
+			k_sem_give(&i2c_info->i2c_transfer_sem);
+			return DRIVER_ERROR_UNSUPPORTED;
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.c:260]: -EINVAL is fine instead of this particular define

File: drivers/i2c/i2c_andes_atciic100.c [hunk 6]
```diff
@@ -0,0 +1,983 @@
...
+		}
+	}
+
+	k_sem_give(&i2c_info->i2c_transfer_sem);
+
+	if (status) {
+		return status;
+	}
+
+	return DRIVER_OK;
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.c:270]: 0 is the ok return in Zephyr generally, so 0 here is fine

File: drivers/i2c/i2c_andes_atciic100.h [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+#define I2C_STAT(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_STAT)
+#define I2C_ADDR(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_ADDR)
+#define I2C_CMD(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_CMD)
+#define I2C_SET(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_SET)
+#define I2C_DATA(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_DATA)
+#define I2C_CTRL(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_CTRL)
+
+#define MEMSET(s, c, n)         __builtin_memset((s), (c), (n))
+#define MEMCPY(des, src, n)     __builtin_memcpy((des), (src), (n))
+#define INWORD(x)     sys_read32(x)
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.h:57]: Seems like an unnecessary wrapper

File: drivers/i2c/i2c_andes_atciic100.h [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+#define I2C_CFG(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_CFG)
+#define I2C_INTE(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_INTE)
+#define I2C_STAT(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_STAT)
+#define I2C_ADDR(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_ADDR)
+#define I2C_CMD(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_CMD)
+#define I2C_SET(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_SET)
+#define I2C_DATA(dev) (DEV_I2C_CFG(dev)->i2c_base_addr + REG_DATA)
+#define I2C_CTRL(dev)  (DEV_I2C_CFG(dev)->i2c_base_addr + REG_CTRL)
+
+#define MEMSET(s, c, n)         __builtin_memset((s), (c), (n))
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.h:55]: Seems like an unnecessary wrapper around memset/memcpy

File: drivers/i2c/i2c_andes_atciic100.h [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+					(SETUP_T_SCL_RATIO_FAST << 13)	| \
+					(SETUP_T_SCLHI_FAST << 4))
+
+#define SETUP_SPEED_FAST_PLUS		((SETUP_T_SUDAT_FAST_P << 24)	| \
+					(SETUP_T_SP_FAST_P  << 21)	| \
+					(SETUP_T_HDDAT_FAST_P << 16)	| \
+					(SETUP_T_SCL_RATIO_FAST_P << 13)| \
+					(SETUP_T_SCLHI_FAST_P << 4))
+
+#define DRIVER_OK			0
```
Comment (teburd) [inline drivers/i2c/i2c_andes_atciic100.h:208]: Use -errno values and avoid redefining your own errors

**Review by teburd — CHANGES_REQUESTED** [review]
Thank you for your contribution!

There's some issues I'd like to see resolved before I look closer at the logic, I've commented on a few spots but the ideas should be generally taken into account.

**Comment by jimmyzhe** [comment]
Thanks for your review, your comment is very helpful to improve this driver, but unfortunately, I found some significant logic errors in this driver, my co-worker will release the correct version of this driver in the future.

I'll close this PR, thanks again for your comment.

**Pull request closed without merging**